server/src/main/java/io/spine/server/delivery/CatchUpStation.java

```diff
@@ -0,0 +1,252 @@
+                                dispatchToCatchUp.put(dispatchingId, message);
+                                conveyor.keepForLonger(message, HOW_LONG_TO_KEEP);
+                            } else {
+                                conveyor.remove(message);
+                            }
+                        } else if (message.getStatus() == TO_DELIVER
+                                && dispatchToCatchUp.containsKey(dispatchingId)) {
+                            conveyor.remove(message);
+                        }
+                    }
+                }
+            }
+        }
+
+        return dispatch(conveyor, dispatchToCatchUp);
+    }
+
+    private Result dispatch(Conveyor conveyor, Map<DispatchingId, InboxMessage> dispatchToCatchUp) {
+        if (!dispatchToCatchUp.isEmpty()) {
+            List<InboxMessage> messages = new ArrayList<>(dispatchToCatchUp.values());
+            messages.sort(COMPARATOR);
+
+            DeliveryErrors errors = action.executeFor(messages);
+            conveyor.markDelivered(messages);
+            Result result = new Result(dispatchToCatchUp.size(), errors);
+            return result;
+        }
+        return emptyResult();
+    }
+
+    /**
+     * Tells whether the job matched the passed {@code InboxMessage}.
+     *
+     * <p>To match, two conditions must be met:
+     *
+     * <ol>
+     *     <li>the target entity type of the job and the message must be the same;</li>
+     *
+     *     <li>the identifier of the message target must be included into the list of the
+     *     identifiers specified in the job OR the job matches all the targets of the entity type.
+     *     </li>
+     * </ol>
+     *
+     * @param job
+     *         the catch-up job
+     * @param message
+     *         the message to match to the job
+     * @return {@code true} if the message matches the job, {@code false} otherwise
+     */
+    @VisibleForTesting
+    static boolean matches(CatchUp job, InboxMessage message) {
+        String expectedProjectionType = job.getId()
+                                           .getProjectionType();
+        InboxId targetInbox = message.getInboxId();
+        String actualTargetType = targetInbox.getTypeUrl();
+        if (!expectedProjectionType.equals(actualTargetType)) {
+            return false;
+        }
+        List<Any> targets = job.getRequest()
+                               .getTargetList();
+        if (targets.isEmpty()) {
+            return true;
+        }
+        Any rawEntityId = targetInbox.getEntityId()
+                                     .getId();
+        return targets.stream()
+                      .anyMatch((t) -> t.equals(rawEntityId));
+    }
+
+    /**
+     * The comparator which sorts the messages chronologically, but ensures that if there is
+     * a {@link CatchUpStarted} event in the sorted batch, it goes on top.
+     */
+    private static final class CatchUpMessageComparator
```

Comment:
Please extract to upper level.

Reply:
I have tried to reduce the number of classes in the `.delivery` package as it's already huge. That's why it's here.

Reply:
I think we'll need to revisit the `delivery` package soon. Presumably, when we get down to working on new Buses. The `process()` method in this class really needs refactoring, and we need a Method Object there. This Method Object would be really linked with the `CatchUpStation` class, while this Comparator is a more separate thing. It can be easier understood being at the top level of the package. It won't hurt. We have many `CatchUp`-classes there. One more there, won't hurt if the price for this would be the `process(` method which is easier to understand.